<launch>
  <!-- RPLIDAR S3 -->
  <include file="$(find rplidar_ros)/launch/rplidar_s3.launch"/>

  <!-- EDIT x y z (m) and yaw (rad) to your real mount -->
  <node pkg="tf" type="static_transform_publisher" name="base_to_laser"
        args="-0.3942 0.5843 0.8500  -1.5381 0 0  base_link laser 50"/>



  <!-- 2D LiDAR odometry -->
<node pkg="laser_scan_matcher" type="laser_scan_matcher_node" name="lsm" output="screen">
  <param name="base_frame"  value="base_link"/>
  <param name="fixed_frame" value="odom"/>         <!-- publishes odom->base_link -->
  <param name="do_publish_tf" value="true"/>

  <!-- Turn OFF external predictions so we stop TF lookups for odom -->
  <param name="use_odom" value="false"/>
  <param name="use_imu"  value="false"/>
  <param name="use_vel"  value="false"/>

  <!-- (Optional) keep defaults, or start conservative: -->
  <param name="max_iterations" value="10"/>
  <remap from="scan" to="/scan"/>
</node>

<node pkg="rosserial_python" type="serial_node.py" name="rosserial" output="screen">
  <param name="port" value="/dev/ttyUSB1"/>
  <param name="baud" value="57600"/>
</node>

  <!-- Your perception & controller -->
<!-- Perception -->
  <node pkg="ransacbug2" type="perception.py" name="perception" output="screen">
    <param name="side" value="right"/>
    <param name="scan_topic" value="/scan"/>
    <param name="sector_center_deg" value="-90.0"/>
    <param name="sector_width_deg"  value="70.0"/>
    <param name="quality_min" value="5.0"/>
    <param name="min_quality_keep" value="3.0"/>
    <param name="front_stop" value="0.55"/>
    <param name="ema_alpha_angle" value="0.3"/>
    <param name="ema_alpha_dist"  value="0.3"/>
    <param name="marker_frame" value="laser"/>
  </node>

<node pkg="ransacbug2" type="bug2.py" name="wall_follower" output="screen">
    <param name="side" value="right"/>
    <param name="d_ref" value="0.50"/>
    <param name="right_overhang" value="0.12"/>   <!-- 先按12cm估，量准后再调 -->
    <param name="band_d" value="0.05"/>
    <param name="k_dist" value="0.60"/>
    <param name="vx_nom" value="0.10"/>
    <param name="vy_max" value="0.35"/>
    <param name="w_max"  value="0.0"/>

    <!-- 安全约束 -->
    <param name="vy_toward_max"  value="0.08"/>   <!-- 朝墙更慢更安全 -->
    <param name="no_toward_band" value="0.10"/>   <!-- 离目标≤10cm 绝不朝墙 -->
    <param name="brake_margin"     value="0.08"/>
    <param name="emergency_margin" value="0.16"/>
    <param name="emergency_time"   value="0.45"/>
    <param name="toward_allow_if_far" value="0.10"/>
    <param name="start_out_bias"   value="0.03"/>
</node>



  <!-- Bag recorder (split to 2GB chunks + LZ4 compression) -->
<node pkg="rosbag" type="record" name="bag" output="screen"
      args="--lz4 --split --size=2048 -O /home/chunxue/bags/wall_1/run
            /scan /odom /pose2D /cmd_vel
            /wall_angle /wall_has /obstacle_ahead
            /tf /tf_static /rosout_agg"/>

</launch>
